torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+            types.push_back(NoneType::get());
+          } else if (t->kind() == NumberType::Kind) {
+            types.push_back(IntType::get());
+            types.push_back(FloatType::get());
+            types.push_back(ComplexType::get());
+          } else {
+            types.push_back(t);
           }
+        };
+
+    auto dedup = [&](std::vector<TypePtr>& types) {
+      std::sort(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) {
+            if (t1->kind() != t1->kind()) {
+              return t1->kind() < t2->kind();
+            }
+            return t1->str() < t2->str();
+          });
+
+      auto new_end = std::unique(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) { return *t1 == *t2; });
+
+      types.erase(new_end, types.end());
+    };
+
+    std::vector<TypePtr> lhs_types;
+    std::vector<TypePtr> rhs_types;
+
+    gather_types(obj, lhs_types, /*is_lhs=*/true);
+    gather_types(classinfo, rhs_types, /*is_lhs=*/false);
+
+    RefinementSet refinement;
+
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    if (obj.kind() == TK_VAR) {
+      std::string ident = Var(obj).name().name();
+
+      std::vector<Refinement> true_refinements;
+      std::vector<Refinement> false_refinements;
+
+      std::vector<TypePtr> isinstance_types;
+      std::vector<TypePtr> not_isinstance_types;
+
+      // We can discard any rhs types that we know statically would be
+      // impossible. For example, if we had:
+      //
+      //    def fn(x: Optional[str]):
+      //        if isinstance(x, (List[str], str, int)):
+      //            ...
+      //
+      // then `x` would be `str` in the true branch and `None` in the
+      // false branch, not `(List[str], str, int)` in the true branch
+      // and `None` in the false branch
+      for (TypePtr lhs_type : lhs_types) {
+        if (lhs_type == AnyType::get()) {
+          isinstance_types.insert(
+              isinstance_types.end(), rhs_types.begin(), rhs_types.end());
+          break;
         }
-        return false;
-      }
-      bool maybeOfKind(TypeKind kind, const TypePtr& actual_type) {
-        if (actual_type->kind() == AnyType::Kind) {
-          return true;
-        }
-        if (const auto optional_type = actual_type->cast<OptionalType>()) {
-          return optional_type->getElementType()->kind() == kind;
+        if (std::any_of(
+                rhs_types.begin(), rhs_types.end(), [&](TypePtr rhs_type) {
+                  return lhs_type->isSubtypeOf(rhs_type);
+                })) {
+          isinstance_types.push_back(lhs_type);
+        } else {
+          not_isinstance_types.push_back(lhs_type);
         }
-        if (const auto union_type = actual_type->cast<UnionType>()) {
+      }
+
+      // Get a single type for the true and false branches
+      unified_true = isinstance_types.empty() ? nullptr
```

Comment:
nit: 
`if (unified_true)` -> `if (isinstance_types.empty())` directly use predicate instead of introducing unnecessary temporary `nullptr`